tests/system/utils.py

```diff
@@ -39,3 +40,122 @@ def get_resource(resource):
+        fault_domain = config.get('fault_domain', {})
+
+        # Extract zone/region
+        self.zone = fault_domain.get('zone', {}).get('name', 'default')
+        self.region = fault_domain.get('region', {}).get('name', 'default')
+
+
+def get_cluster_local_domain():
+    """Contacts the DC/OS mesos master and returns it's faultDomain configuration (aka "local domain")
+    """
+    master = mesos.get_master()
+    return FaultDomain(master.state().get('domain', {}))
+
+
+def get_cluster_agent_domains():
+    """Returns a dictionary with the slave IDs in the cluster and their corresponding
+       fault domain information
+    """
+    agent_domains = {}
+
+    # Populate agent_domains with the ID and the corresponding domain for each slave
+    master = mesos.get_master()
+    for slave in master.slaves():
+        agent_domains[slave['id']] = FaultDomain(slave._short_state.get('domain', None))
+    return agent_domains
+
+
+def get_all_cluster_regions():
+    """Returns a dictionary with all the regions and their zones in the cluster
+    """
+    domain_regions = {}
+
+    # Populate all the domain zones and regions found in the cluster
+    for domain in get_cluster_agent_domains().values():
+        if domain.region not in domain_regions:
+            domain_regions[domain.region] = []
+        if domain.zone not in domain_regions[domain.region]:
+            domain_regions[domain.region].append(domain.zone)
+
+    # Return dictionary
+    return domain_regions
+
+
+def get_biggest_cluster_region():
+    """Returns a tuple with the name of the region with the most zones in the
+       cluster and a list with the actual zones in it
+    """
+    biggest_region = None
+    biggest_region_zones = []
+
+    for (region, zones) in get_all_cluster_regions().items():
+        if len(zones) > len(biggest_region_zones):
+            biggest_region_zones = zones
+            biggest_region = region
+
+    return (biggest_region, biggest_region_zones)
+
+
+def get_app_domains(app):
+    """Returns a list of all te fault domains used by the tasks of the specified app
+    """
+    tasks = app.get('tasks', [])
+    agent_domains = get_cluster_agent_domains()
+
+    assert len(tasks) > 0, "App %s did not launch any tasks on mesos" % (app['id'],)
+
+    # Collect the FaultDomain objects from all the agents where the tasks are running
+    domains = set()
+    for task in tasks:
+        domains.append(agent_domains[task['slaveId']])
+
+    return domains
+
+
+def get_used_regions_and_zones(domains):
+    """Returns tuple with the sets of the used regions and zones from the given list
+       of FaultDomain object instances (ex. obtained by get_app_domains())"""
+    used_regions = set()
+    used_zones = set()
+
+    for domain in domains:
+        used_regions.add(domain.region)
+        used_zones.add(domain.zone)
+
+    return (used_regions, used_zones)
+
+
+def count_agents_in_faultdomains(regions=None, zones=None):
```

Comment:
Are we really checking for the fault domain as one can pass regions and zones.

Reply:
This is Python 3.5; why not use optional typing?

Generally I prefer methods that are strict in what they accept. Can we push the concern to make sure they're always iterable upstream? IE passing a string to this method should be an error, not something we just fix and sweep under the rug.

Also, does it make sense to have it accept a python set, and default regions / zones default to an empty set, rather than `None`?